src/node_file.cc

```diff
@@ -654,17 +768,33 @@ static void ReadLink(const FunctionCallbackInfo<Value>& args) {
 
   if (args.Length() < 1)
     return TYPE_ERROR("path required");
-  if (!args[0]->IsString())
-    return TYPE_ERROR("path must be a string");
 
-  node::Utf8Value path(env->isolate(), args[0]);
+  Path path(env->isolate(), args[0]);
+  if (path.fail())
+    return TYPE_ERROR("path must be a string or Buffer");
+
+  enum encoding encoding = UTF8;
+  Local<Value> callback = Null(env->isolate());
+  if (args.Length() == 2) {
+    if (args[1]->IsObject()) {
+      callback = args[1];
+    } else if (args[1]->IsString()) {
+      encoding = ParseEncoding(env->isolate(), args[1], UTF8);
+    }
+  } else if (args.Length() == 3) {
+    callback = args[2];
+    encoding = ParseEncoding(env->isolate(), args[1], UTF8);
+  }
```

Comment:
Is there a reason you can't modify lib/fs.js so that the binding layer always gets called with the same number (and kinds) of arguments?